File: lollys-lotto-rust-sdk/src/instructions/cranks/crank_transfer_winning_amount_to_user_rewards_vault.rs
use crate::instructions::*;
use lollys_lotto::{
    instruction::CrankTransferWinningAmountToUserRewardsVault, state::LottoTicketNumbers,
};

pub fn crank_transfer_winning_amount_to_user_rewards_vault(
    round: u64,
    winning_numbers: LottoTicketNumbers,
    number_of_tickets_with_duplicate_numbers: u32,
    authority: &Pubkey,
    user: &Pubkey,
    user_metadata: &Pubkey,
    user_rewards_vault: &Pubkey,
    lotto_game: &Pubkey,
    lotto_game_vault_signer: &Pubkey,
    lotto_game_vault: &Pubkey,
    lotto_ticket: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = CrankTransferWinningAmountToUserRewardsVault {
        round,
        winning_numbers,
        number_of_tickets_with_duplicate_numbers,
    }
    .data();
    let accounts = lollys_lotto::accounts::CrankTransferWinningAmountToUserRewardsVault {
        authority: *authority,
        user: *user,
        user_metadata: *user_metadata,
        user_rewards_vault: *user_rewards_vault,
        lotto_game: *lotto_game,
        lotto_game_vault_signer: *lotto_game_vault_signer,
        lotto_game_vault: *lotto_game_vault,
        lotto_ticket: *lotto_ticket,
        event_emitter: *event_emitter,
        token_program: token::ID,
    }
    .to_account_metas(None);
    Instruction {
        program_id: lollys_lotto::ID,
        accounts,
        data,
    }
}

File: lollys-lotto-ts-sdk/src/codegen/instructions/crankTransferWinningAmountToUserRewardsVault.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CrankTransferWinningAmountToUserRewardsVaultArgs {
  round: BN
  winningNumbers: types.LottoTicketNumbersFields
  numberOfTicketsWithDuplicateNumbers: number
}

export interface CrankTransferWinningAmountToUserRewardsVaultAccounts {
  authority: PublicKey
  lottoGame: PublicKey
  lottoGameVaultSigner: PublicKey
  lottoGameVault: PublicKey
  user: PublicKey
  userMetadata: PublicKey
  userRewardsVault: PublicKey
  lottoTicket: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.u64("round"),
  types.LottoTicketNumbers.layout("winningNumbers"),
  borsh.u32("numberOfTicketsWithDuplicateNumbers"),
])

export function crankTransferWinningAmountToUserRewardsVault(
  args: CrankTransferWinningAmountToUserRewardsVaultArgs,
  accounts: CrankTransferWinningAmountToUserRewardsVaultAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: true, isWritable: false },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: true },
    {
      pubkey: accounts.lottoGameVaultSigner,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.lottoGameVault, isSigner: false, isWritable: true },
    { pubkey: accounts.user, isSigner: false, isWritable: false },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    { pubkey: accounts.userRewardsVault, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoTicket, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([101, 178, 55, 235, 11, 59, 135, 113])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      round: args.round,
      winningNumbers: types.LottoTicketNumbers.toEncodable(args.winningNumbers),
      numberOfTicketsWithDuplicateNumbers:
        args.numberOfTicketsWithDuplicateNumbers,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).subarray(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
